File: leetcode/[array]448. Find All Numbers Disappeared in an Array.cpp
/**
 * [ 배열 ]
 * 문제: 448. Find All Numbers Disappeared in an Array
 *
 * https://leetcode.com/problems/find-all-numbers-disappeared-in-an-array/submissions/
 */
#include <iostream>
#include <cstdio>
#include <vector>
#include <queue>
#include <algorithm>
#include <cmath>
#include <list>
using namespace std;

vector<int> findDisappearedNumbers(vector<int>& nums) {
    // you may assume the returned list does not count as extra space.
    vector<int> answer;
    int size = nums.size();

    for (int i = 0; i < size; i++) {

        // 이미 방문한 경우는 "음수"처리해버리기 때문에, 음수인 경우는 Index 탐색을 위해 양수로 돌려준다.
        int tmp = nums[i] > 0 ? nums[i] : -nums[i];

        // 방문처리: "음수"로 만듬으로 방문 처리
        // 여기서 굳이 -1 하는 경우는 => 실제 size는 1~n 이지만, index 는 0 ~ n-1이기 때문에
        nums[tmp-1] = nums[tmp-1] > 0 ? -nums[tmp-1] : nums[tmp-1];

    }

    for (int i = 0; i < size; i++) {

        // 양수: vector의 숫자가 없기 때문에 방문하지 못함을 의미함
        if (nums[i] > 0) {
            // index보다 +1된 값이 진짜 숫자임
            answer.push_back(i+1);
        }
    }

    return answer;
}

int main() {


    // 결과
    // Runtime: 52 ms, faster than 83.14% of C++ online submissions for Find All Numbers Disappeared in an Array.
    // Memory Usage: 33.5 MB, less than 94.57% of C++ online submissions for Find All Numbers Disappeared in an Array.

    vector<int> nums;
    nums.push_back(3);
    nums.push_back(4);
    nums.push_back(2);
    nums.push_back(7);
    nums.push_back(8);
    nums.push_back(3);
    nums.push_back(1);
    nums.push_back(2);


    vector<int> answer = findDisappearedNumbers(nums);
    for (int i = 0; i < answer.size(); i++) {
        cout << answer[i] << " ";
    }
    cout << endl;

    return 0;
}

File: leetcode/[array]448. Find All Numbers Disappeared in an Array.go
package main

import (
	"fmt"
)

func findDisappearedNumbers(nums []int) []int {
	var answer []int
	size := len(nums)

	for i := 0; i < size; i++ {

		// 방문할 수 있는 숫자 탐색
		var tmp int
		if nums[i] > 0 {
			tmp = nums[i]
		} else {
			// X -1 경우:  이미 방문한 노드일 경우 음수처리가 되어있기 때문에 양수로 변경시킨다!
			tmp = nums[i]*-1
		}

		// 방문처리: 음수로 표기하면서 방문했음을 처리한다.
		// -1하는이유: 실제 size는 1~n 이지만, index 는 0 ~ n-1이기 때문에
		if nums[tmp-1] > 0 {
			nums[tmp-1] = nums[tmp-1]*-1
		}

	}

	// 정답 찾기
	for i := 0; i < size; i++ {

		// 양수: 한번도 방문하지못함 -> 숫자가 없었기 때문에 방문을 못한것
		if nums[i] > 0 {
			// +1 이유: 실제 size는 1~n 이지만, index 는 0 ~ n-1이기 때문에
			answer = append(answer, i+1)
		}
	}

	return answer
}

func main()  {

	// 코드 결과
	// Runtime: 52 ms, faster than 89.23% of Go online submissions for Find All Numbers Disappeared in an Array.
	// Memory Usage: 7 MB, less than 79.23% of Go online submissions for Find All Numbers Disappeared in an Array.

	nums := []int{4,3,2,7,8,2,3,1}
	fmt.Printf("%d\n", findDisappearedNumbers(nums))
}
